go/pkg/reader/reader.go

```diff
@@ -95,3 +99,69 @@ func (fio *fileSeeker) Initialize() error {
 	fio.initialized = true
 	return nil
 }
+
+type compressedSeeker struct {
+	fs   ReadSeeker
+	encd *zstd.Encoder
+	// This keeps the compressed data
+	buf *bytes.Buffer
+}
+
+func NewCompressedFileSeeker(path string, buffsize int) (ReadSeeker, error) {
+	return NewCompressedSeeker(NewFileReadSeeker(path, buffsize))
+}
+
+func NewCompressedSeeker(fs ReadSeeker) (ReadSeeker, error) {
+	if _, ok := fs.(*compressedSeeker); ok {
+		return nil, errors.New("Trying to double compress files.")
+	}
+
+	buf := bytes.NewBuffer(nil)
+	encd, err := zstd.NewWriter(buf)
+	return &compressedSeeker{
+		fs:   fs,
+		encd: encd,
+		buf:  buf,
+	}, err
+}
+
+func (cfs *compressedSeeker) Read(p []byte) (int, error) {
+	var n int
+	var errR, errW error
+	for cfs.buf.Len() < len(p) && errR == nil && errW == nil {
+		// Read is allowed to use the entity of p as a scratchpad.
+		n, errR = cfs.fs.Read(p)
+		// errW must be non-nil if written bytes != from n.
+		_, errW = cfs.encd.Write(p[:n])
+	}
+	m, err := cfs.buf.Read(p)
+
+	var retErr error
+	if errR == io.EOF {
+		cfs.encd.Close()
+	} else if errR != nil {
+		retErr = errR
+	}
+	if retErr == nil && errW != nil {
+		retErr = errW
+	}
+	if retErr == nil {
+		retErr = err
+	}
+	return m, err
+}
+
+func (cfs *compressedSeeker) SeekOffset(offset int64) error {
+	cfs.buf.Reset()
+	err := cfs.encd.Close()
+	cfs.encd.Reset(cfs.buf)
```

Comment:
Why not return err immediately if it is not nil?

Reply:
Hah, I had a pending comment reply on #230. Basically it'd allow you to re-use a reader - but that's not what we are planning to do anymore.